File: web/src/services/types/prompt.ts
export type PromptSource = 'official' | 'workspace' | 'personal';
export type PromptType = 'text' | 'chat';
export type PromptOptimizerGoal = 'general' | 'reliable' | 'structured' | 'deep';
export type PromptOptimizerVariant = 'safe' | 'balanced' | 'advanced';

export interface PromptPlaygroundMessage {
  role: 'system' | 'user' | 'assistant';
  content: string;
}

export interface PromptVersionPayload {
  prompt_type: PromptType;
  content: string | Array<{ role: 'system' | 'user' | 'assistant'; content: string }>;
  config?: Record<string, unknown>;
  labels?: string[];
  commit_message?: string | null;
}

export interface CreatePromptRequest {
  workspace_id: string;
  source: Exclude<PromptSource, 'official'>;
  name: string;
  slug?: string;
  description?: string | null;
  locale?: string;
  category?: string | null;
  tags?: string[];
  initial_version: PromptVersionPayload;
}

export interface UpdatePromptRequest {
  name?: string;
  description?: string | null;
  locale?: string;
  category?: string | null;
  tags?: string[];
  source?: Exclude<PromptSource, 'official'>;
}

export interface SetPromptLabelsRequest {
  version: number;
  labels: string[];
}

export interface PromptVersion {
  id: string;
  version: number;
  prompt_type: PromptType;
  content: string | Array<{ role: 'system' | 'user' | 'assistant'; content: string }>;
  config: Record<string, unknown>;
  labels: string[];
  commit_message?: string | null;
  created_by?: string | null;
  created_at: string;
  updated_at: string;
}

export interface PromptSummary {
  id: string;
  workspace_id?: string | null;
  owner_account_id?: string | null;
  source: PromptSource;
  name: string;
  slug: string;
  description?: string | null;
  locale: string;
  category?: string | null;
  tags: string[];
  latest_version: number;
  latest_labels: string[];
  latest_prompt_type: PromptType;
  is_owned: boolean;
  created_at: string;
  updated_at: string;
}

export interface PromptDetail extends PromptSummary {
  versions: PromptVersion[];
}

export interface PromptPickerSelection {
  prompt: PromptSummary;
  version: PromptVersion;
  reference:
    | {
        mode: 'label';
        label: string;
      }
    | {
        mode: 'version';
        version: number;
      };
}

export interface PromptListParams {
  page?: number;
  limit?: number;
  keyword?: string;
  workspace_id?: string;
  locale?: string;
  source?: PromptSource;
  category?: string;
}

export interface PromptList {
  data: PromptSummary[];
  has_more: boolean;
  limit: number;
  page: number;
  total: number;
}

export interface PromptOptimizeRequest {
  raw_prompt: string;
  goal?: PromptOptimizerGoal;
  preserve_variables?: boolean;
  provider?: string;
  model?: string;
  prompt_id?: string;
}

export interface PromptOptimizeResult {
  goal: PromptOptimizerGoal;
  preserve_variables: boolean;
  detected_variables: string[];
  run_id: string;
  output: string;
  variants: Record<PromptOptimizerVariant, string>;
}

export interface PromptPlaygroundRequest {
  prompt: string;
  messages?: PromptPlaygroundMessage[];
  input?: string;
  variables?: Record<string, string>;
  provider?: string;
  model?: string;
}

export interface PromptUsageReference {
  agent_id: string;
  agent_name: string;
  workflow_id: string;
  node_id: string;
  node_title: string;
  reference_mode?: string;
  label?: string | null;
  version?: number | null;
  updated_at: string;
}

export interface PromptUsageRecentRun {
  workflow_run_id?: string | null;
  agent_id: string;
  agent_name: string;
  node_id: string;
  node_title: string;
  status: string;
  prompt_version?: number | null;
  requested_label?: string | null;
  requested_version?: number | null;
  created_at: string;
  finished_at?: string | null;
  elapsed_time: number;
}

export interface PromptUsageSummary {
  linked_agents_count: number;
  linked_nodes_count: number;
  total_run_count: number;
  last_run_at?: string | null;
  version_metrics: Array<{
    version: number;
    run_count: number;
    last_run_at?: string | null;
  }>;
  label_metrics: Array<{
    label: string;
    run_count: number;
    last_run_at?: string | null;
  }>;
  references: PromptUsageReference[];
  recent_runs: PromptUsageRecentRun[];
}

export interface PromptOptimizationRun {
  id: string;
  prompt_id?: string | null;
  goal: PromptOptimizerGoal;
  provider?: string | null;
  model?: string | null;
  preserve_variables: boolean;
  detected_variables: string[];
  raw_prompt: string;
  output: string;
  variants: Record<PromptOptimizerVariant, string>;
  adopted_variant?: PromptOptimizerVariant | null;
  adopted_prompt_version_id?: string | null;
  adopted_at?: string | null;
  created_at: string;
  updated_at: string;
}

export interface PromptOptimizationRunList {
  data: PromptOptimizationRun[];
  has_more: boolean;
  limit: number;
  page: number;
  total: number;
}

export interface PromptOptimizationRunListParams {
  page?: number;
  limit?: number;
}

export interface AdoptPromptOptimizationRunRequest {
  variant: PromptOptimizerVariant;
  commit_message?: string | null;
}

File: api/internal/modules/prompts/model/prompt.go
package model

import (
	"time"

	"github.com/google/uuid"
	"gorm.io/datatypes"
	"gorm.io/gorm"
)

type PromptSource string

const (
	PromptSourceOfficial  PromptSource = "official"
	PromptSourceWorkspace PromptSource = "workspace"
	PromptSourcePersonal  PromptSource = "personal"
)

type PromptType string

const (
	PromptTypeText PromptType = "text"
	PromptTypeChat PromptType = "chat"
)

type Prompt struct {
	ID             string          `gorm:"type:uuid;primaryKey" json:"id"`
	OrganizationID *string         `gorm:"type:uuid;index:idx_app_prompts_org_workspace_source,priority:1" json:"organization_id,omitempty"`
	WorkspaceID    *string         `gorm:"type:uuid;index:idx_app_prompts_org_workspace_source,priority:2" json:"workspace_id,omitempty"`
	OwnerAccountID *string         `gorm:"type:uuid;index" json:"owner_account_id,omitempty"`
	Source         PromptSource    `gorm:"type:varchar(32);not null;index:idx_app_prompts_org_workspace_source,priority:3" json:"source"`
	Name           string          `gorm:"type:varchar(255);not null" json:"name"`
	Slug           string          `gorm:"type:varchar(255);not null;index" json:"slug"`
	Description    *string         `gorm:"type:text" json:"description,omitempty"`
	Locale         string          `gorm:"type:varchar(32);not null;default:'zh-Hans';index" json:"locale"`
	Category       *string         `gorm:"type:varchar(128);index" json:"category,omitempty"`
	Tags           []string        `gorm:"type:jsonb;serializer:json;not null;default:'[]'" json:"tags"`
	LatestVersion  int             `gorm:"not null;default:1" json:"latest_version"`
	CreatedAt      time.Time       `json:"created_at"`
	UpdatedAt      time.Time       `json:"updated_at"`
	Versions       []PromptVersion `gorm:"foreignKey:PromptID;references:ID" json:"versions,omitempty"`
}

func (Prompt) TableName() string {
	return "app_prompts"
}

func (p *Prompt) BeforeCreate(tx *gorm.DB) error {
	if p.ID == "" {
		p.ID = uuid.New().String()
	}
	now := time.Now()
	if p.CreatedAt.IsZero() {
		p.CreatedAt = now
	}
	if p.UpdatedAt.IsZero() {
		p.UpdatedAt = now
	}
	return nil
}

type PromptVersion struct {
	ID            string         `gorm:"type:uuid;primaryKey" json:"id"`
	PromptID      string         `gorm:"type:uuid;not null;index:idx_app_prompt_versions_prompt_version,priority:1" json:"prompt_id"`
	Version       int            `gorm:"not null;index:idx_app_prompt_versions_prompt_version,priority:2" json:"version"`
	PromptType    PromptType     `gorm:"type:varchar(16);not null" json:"prompt_type"`
	Content       datatypes.JSON `gorm:"type:jsonb;not null" json:"content"`
	Config        datatypes.JSON `gorm:"type:jsonb;not null;default:'{}'" json:"config"`
	Labels        []string       `gorm:"type:jsonb;serializer:json;not null;default:'[]'" json:"labels"`
	CommitMessage *string        `gorm:"type:text" json:"commit_message,omitempty"`
	CreatedBy     *string        `gorm:"type:uuid" json:"created_by,omitempty"`
	CreatedAt     time.Time      `json:"created_at"`
	UpdatedAt     time.Time      `json:"updated_at"`
}

func (PromptVersion) TableName() string {
	return "app_prompt_versions"
}

func (p *PromptVersion) BeforeCreate(tx *gorm.DB) error {
	if p.ID == "" {
		p.ID = uuid.New().String()
	}
	now := time.Now()
	if p.CreatedAt.IsZero() {
		p.CreatedAt = now
	}
	if p.UpdatedAt.IsZero() {
		p.UpdatedAt = now
	}
	return nil
}

type PromptOptimizationRun struct {
	ID                     string     `gorm:"type:uuid;primaryKey" json:"id"`
	OrganizationID         string     `gorm:"type:uuid;not null;index:idx_prompt_opt_runs_org_workspace_created,priority:1" json:"organization_id"`
	WorkspaceID            *string    `gorm:"type:uuid;index:idx_prompt_opt_runs_org_workspace_created,priority:2" json:"workspace_id,omitempty"`
	PromptID               *string    `gorm:"type:uuid;index:idx_prompt_opt_runs_account_prompt_created,priority:2" json:"prompt_id,omitempty"`
	AccountID              string     `gorm:"type:uuid;not null;index:idx_prompt_opt_runs_account_prompt_created,priority:1" json:"account_id"`
	Goal                   string     `gorm:"type:varchar(32);not null;default:'general'" json:"goal"`
	Provider               *string    `gorm:"type:varchar(128)" json:"provider,omitempty"`
	Model                  *string    `gorm:"type:varchar(255)" json:"model,omitempty"`
	PreserveVariables      bool       `gorm:"not null;default:true" json:"preserve_variables"`
	DetectedVariables      []string   `gorm:"type:jsonb;serializer:json;not null;default:'[]'" json:"detected_variables"`
	RawPrompt              string     `gorm:"type:text;not null" json:"raw_prompt"`
	SafeOutput             string     `gorm:"type:text;not null" json:"safe_output"`
	BalancedOutput         string     `gorm:"type:text;not null" json:"balanced_output"`
	AdvancedOutput         string     `gorm:"type:text;not null" json:"advanced_output"`
	AdoptedVariant         *string    `gorm:"type:varchar(16)" json:"adopted_variant,omitempty"`
	AdoptedPromptVersionID *string    `gorm:"type:uuid" json:"adopted_prompt_version_id,omitempty"`
	AdoptedAt              *time.Time `json:"adopted_at,omitempty"`
	CreatedAt              time.Time  `gorm:"index:idx_prompt_opt_runs_account_prompt_created,priority:3;index:idx_prompt_opt_runs_org_workspace_created,priority:3" json:"created_at"`
	UpdatedAt              time.Time  `json:"updated_at"`
}

func (PromptOptimizationRun) TableName() string {
	return "app_prompt_optimization_runs"
}

func (r *PromptOptimizationRun) BeforeCreate(tx *gorm.DB) error {
	if r.ID == "" {
		r.ID = uuid.New().String()
	}
	now := time.Now()
	if r.CreatedAt.IsZero() {
		r.CreatedAt = now
	}
	if r.UpdatedAt.IsZero() {
		r.UpdatedAt = now
	}
	return nil
}
